### PR TITLE
[FH #1] GIF MAKER

### PR DESCRIPTION
**Pull request by alekcangp**
### Create the animated Frame from a video file or YouTube.
It can be used to post frames with a short intro/demo video or funny GIFs.

Adjustments: 
* start and duration time for a video clip
* caption on an animated gif
* caption's position
* caption's size
* caption's color
* font style
* frame's button label
* frame's button link

<p align="center">

</p>

Links to a Farcaster posts: [far.quest](https://far.quest/alekcangp/0x96e5d3b4e505126838800f1f46090fba063400bd), [warpcast](https://warpcast.com/alekcangp/0x96e5d3b4)

### CODE REVIEW

**Comment by FTCHD** [comment]
API calls to external services that are not publicly known and vetted is not allowed.

You cannot call external APIs that you own, it's one thing calling the 0x or OpenSea API, and another calling your own API that you host somewhere that is not recognized, not verified, and can change at any time.

In that sense, please move the GIF creation handling inside FrameTrain, ideally client side in the Inspector using a worker. If you are experiencing issues please visit the dedicated channel on Warpcast.

**Comment by alekcangp** [comment]
This API uses **BASH** scripts and dependences from this repo [https://github.com/llimllib/ytgif](https://github.com/llimllib/ytgif) and is deployed on own VPS. Is there any proper way to do it  inside FrameTrain?

**Comment by FTCHD** [comment]
You have to find a way to do it client-side in your `Inspector.tsx`, either through a service worker or otherwise. See the PDF template that renders the PDF to images client-side. 

You can install npm packages that do this client side, but not use external API endpoints that are not validated and well known.

A short google search yielded the following results:
- https://github.com/incubated-geek-cc/video-to-GIF
- https://javascript.plainenglish.io/how-to-convert-a-video-clip-to-a-gif-file-with-client-side-javascript-56575d093191
- https://yahoo.github.io/gifshot/
- https://javascript.plainenglish.io/how-to-build-a-gif-creator-in-browser-3235d274c27f
- https://www.reddit.com/r/programming/comments/1f3tor/gifjs_fast_gif_encoder_using_web_workers/?rdt=61071

You can do a lot of things client side nowadays.

**Comment by alekcangp** [comment]
Done.

**Comment by FTCHD** [comment]
Do users have to upload the video themselves now?

**Comment by alekcangp** [comment]
Yes, added `Upload a video` button and video/gif preview.

File: templates/gifmaker/Inspector.tsx [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+		caption: 'Hello from FrameTrain',
+		fontsize: '30',
+		fontcolor: 'white',
+		y: '20',
+		label: 'Do It',
+		link: 'https://frametra.in',
+	};
+
+
+	//CRATE GIF FROM PARAMS AND PREVIEW
```
Comment (FTCHD) [inline templates/gifmaker/Inspector.tsx:59]: Spelling mistake

File: templates/gifmaker/Inspector.tsx [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	const inputFontColor = useRef<HTMLInputElement>(null);
+	const inputY = useRef<HTMLInputElement>(null);
+	const inputButtonLabel = useRef<HTMLInputElement>(null);
+	const inputButtonLink = useRef<HTMLInputElement>(null);
+
+	const confDefault = {
+		start: '0',
+		duration: '10',
+		caption: 'Hello from FrameTrain',
+		fontsize: '30',
```
Comment (FTCHD) [inline templates/gifmaker/Inspector.tsx:51]: Follow camelCase

File: templates/gifmaker/Inspector.tsx [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+				fontcolor: inputFontColor.current?.value || confDefault.fontcolor,
+				label: inputButtonLabel.current?.value || confDefault.label,
+				link: inputButtonLink.current?.value || confDefault.link,
+			};
+			const ffmpeg = ffmpegRef.current;
+			const ty = file.type.substring(file.type.indexOf('/') + 1);
+			await ffmpeg.writeFile(`input.${ty}`, await fetchFile(file));
+			await ffmpeg.writeFile(
+				'arial.ttf',
+				await fetchFile('https://ipfs.io/ipfs/QmcWFzXah4SAbKruiDjUN9rgSzztDwJU4F9d6qreRZNF1C'),
```
Comment (FTCHD) [inline templates/gifmaker/Inspector.tsx:81]: Fetch font from a CDN or Google Fonts, not IPFS

File: templates/gifmaker/index.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    name: 'GIF Maker',
+    description: 'Create a GIF/Trailer Frame from video.',
+    creatorFid: '416810',
+    creatorName: 'alekcangp',
+    cover,
+    enabled: true,
+    Inspector,
+    functions,
+    initialConfig: {
+        gif: 'https://ipfs.io/ipfs/QmZ3SuHbBcZvyBTsCEWZgYFXBXpUW9TQQLctv6Fr2Eva9U',
```
Comment (FTCHD) [inline templates/gifmaker/index.ts:24]: Same as above, don't use unreadable IPFS links

**Comment by FTCHD** [comment]
Left a few comments in the code above.

Additionally:
1. Please check all spelling mistakes, both code and comments.
2. Follow naming conventions (`camelCase`).
3. Provide clear naming to the fields in the Inspector, it's not clear to me what anything should do here, or what will happen when I type into one of the fields. Compare your Inspector panel with the Inspectors of other templates. Clear field names, hint text below the fields (where relevant), etc. Each field must have a Label, and some of then surely require some extra Hint Text (see the Discourse template, the text under the field).
4. Use the right components. Color is being picked using ColorPicker, not a field where you type "blue".

**Comment by alekcangp** [comment]
Done.

**Comment by FTCHD** [comment]
Please read the documentation and use the integrated Font components. 

https://github.com/FTCHD/frametrain/wiki/%5BSDK%5D-Components#fontfamilypicker

Also, update the image of your Frame in this PR.

**Comment by alekcangp** [comment]
Done.

**Comment by FTCHD** [comment]
### Extra Buttons

There should be no "Preview", "Create Frame" or "Pre-saved configuration" button. 

It's your duty as the template artist to feed the default values into the fields or functions, if they are empty. There's no need for a "Reset" button. What you may refer to as a "Reset" button should be a button to remove the current uploaded video (or upload a new one). See the PDF Template for this handling.

Apart from that, there shouldn't be any buttons, everything is reactive and happens when you upload a video or change a field value. The `update` function is debounced internally so it's not gonna trigger continuously, it's gonna trigger often tho, but that is intended. You shouldn't have to press anything to trigger changes, just make the change.

### Console

What is the purpose of "Console" and can it be removed?

**Comment by alekcangp** [comment]
Done.

**Comment by FTCHD** [comment]
Thank you! To note, unused files in the bucket are deleted every 24hrs to avoid such issues.

**Comment by alekcangp** [comment]
Added possibility to create GIF from YouTube.

**Comment by FTCHD** [comment]
Screenshot 2024-06-29 at 3 13 03 PM

This code right here appears to be dead at best, and very buggy at worst. `onChange` don't trigger on `divs`, and `class` is not a valid property in React. Additionally, the `file` source is always selected, so even changing the current source type does not do anything.

You must've written this on your phone to not get errors about it in your IDE.

**Comment by FTCHD** [comment]
The link you provided shows it working but it's not congruent with the code that I'm seeing here.

**Comment by alekcangp** [comment]
Fixed.

**Comment by FTCHD** [comment]
Awesome, please respond on Warpcast!

**Pull request merged**